lib/AST/ASTMangler.cpp

```diff
@@ -1359,7 +1359,7 @@ void ASTMangler::appendFlatGenericArgs(SubstitutionMap subs) {
   for (auto replacement : subs.getReplacementTypes()) {
     if (replacement->hasArchetype())
       replacement = replacement->mapTypeOutOfContext();
-    appendType(replacement);
+    appendType(replacement->getCanonicalType());
```

Comment:
It would be better to ensure that any SubstitutionMap that ends up here was canonicalized first